001-public-stream-base.md

```diff
@@ -0,0 +1,166 @@
+# Make C++ Stream APIs public
+
+Author: Fedor Indutny <fedor@indutny.com>
+Status: draft
+Date: Thu Dec 17 2015 19:39:46 GMT-0500 (EST)
+
+## Description
+
+Currently we use so called `StreamBase` APIs through the core to implement
+various performance-critical networking protocols with maximum efficiency.
+It is evident that this performance improvement could be beneficial for
+user-land C++ addons too.
+
+I propose to make this APIs public, introducing C++ interface similar to the
+following one (simplified):
+
+## Interface
+
+```C++
+class Stream {
+ public:
+  //
+  // Low-level part
+  //
+
+  // Perform protocol-specific shutdown (EOF)
+  virtual int DoShutdown(ShutdownWrap* req_wrap) = 0;
+
+  // Write data asynchronously
+  virtual int DoWrite(WriteWrap* w,
+                      uv_buf_t* bufs,
+                      size_t count,
+                      uv_stream_t* send_handle) = 0;
+
+  // **(optional)**
+  // Try to write provided data immediately without blocking, if not
+  // possible to do - should return `0` if no data was written, should decrement
+  // `count` and change pointers/length in `bufs` if some data was written.
```

Comment:
Can we include a helper function or whatnot for recalculating the `uv_buf_t**`? We have the logic already in `StreamWrap::DoTryWrite()`, and it'd be a shame for everyone to need to implement basically the exact same on their own.